Product sets: basic pdp (@W-12301851@)

vmarta opened the pull request:
For the 1st round of implementation, we aim to have a basic PDP working that supports a product set. 

However, the original code assumed that there's only a single product on the page. For multiple products, we'll need to tweak how we save the swatches state (variation attributes) into the page url's search parameters. Otherwise, selecting a size, for example, would select that size for all of the child products.

With this PR.. for a **single product**, the page url would look something like this (still the same as before):
`/product/25502228M?color=JJ0NLD0&size=9MD`

For **multiple products**, the url is like this:
`/product/Spring-look-2M?25565826M=color%3DJJGG3XX%26size%3D9LG&25518009M=color%3DJJ9LMXX`
(Shopper has selected some swatches for the child products, whose ids are `25565826M` and `25518009M`)

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

Example of a PDP with product set: navigate to this list http://localhost:3000/global/en-GB/category/womens-outfits

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

vmarta committed:
Product set: render the children products with minimal UI

vmarta committed:
No longer showing breadcrumbs in the child items

vmarta committed:
Merge branch 'feature/product-sets' into product-sets-basic-pdp

vmarta committed:
Individual swatches now have correct `value`

There was a subtle bug where the `value` of the SwatchGroup was passed down to override all the values of the children Swatch components.

vmarta committed:
Merge branch 'feature/product-sets' into product-sets-basic-pdp

vmarta committed:
1st attempt at saving swatches state in the page url

vmarta committed:
Avoid null in the url search params

vmarta committed:
Some refactoring

vmarta committed:
Add-to-cart now works for child items

vmarta committed:
Show accordion for each child item

vmarta committed:
Add todo

vmarta committed:
Rename variable for accuracy

The child product itself is actually not a set, but is considered a set product.

vmarta committed:
Optional argument

vmarta committed:
Some refactoring and fix linting

vmarta committed:
Helper hook for url search params

vmarta committed:
Clean up

vmarta committed:
Rename module for a specific use case (PDP)

vmarta commented on packages/template-retail-react-app/app/components/swatch-group/index.jsx:
```diff
@@ -28,9 +35,7 @@ const SwatchGroup = (props) => {
                     return React.cloneElement(child, {
-                        selected: childValue === value,
-                        key: childValue,
-                        value,
```
Aside: there was a subtle bug with the original code. Now the parent (SwatchGroup) no longer overriding the children's `value`. They should be distinct.

vmarta commented on packages/template-retail-react-app/app/pages/product-detail/index.jsx:
```diff
@@ -155,105 +147,36 @@ const ProductDetail = ({category, product, isLoading}) => {
+                        <Fragment key={childProduct.id}>
+                            <ProductView
+                                product={childProduct}
+                                isSetProduct={true}
```
I added this `isSetProduct` for the ProductView component (and also the related React hooks) because given solely the data for a _child_ product, it is not possible to identify whether the product belongs to a set or not.

bendvc replied:
Have you thought about other product types that we currently don't support that we might in the future. And how using `isSetProduct` as a prop will age? 

Side note: Is `isSetProduct` short for `isProductSetProduct`

vmarta replied:
@bendvc Good questions.. I don't know yet at this time, but there's a separate refactoring ticket for supporting other product types.🤔 However, I'll think about it a bit more at this stage anyways.

vmarta replied:
> Is isSetProduct short for isProductSetProduct

Yes, `isSetProduct` comes from the way SCAPI calls the child products. The data for a product set has `setProducts: [...]`.

vmarta commented on packages/template-retail-react-app/app/hooks/use-variation-params.js:
```diff
@@ -5,25 +5,27 @@
-    const params = new URLSearchParams(search)
-    // Using all the variation attribute id from the array generated above, get
+    const [allParams, productParams] = usePDPSearchParams(product.id)
+    const params = isSetProduct ? productParams : allParams
```
I tried to reuse the existing React hooks. The main difference is for set products, the hooks focus in on a specific subset of the page url params, rather than all of the params.

vmarta committed:
Show/hide add-to-cart buttons accordingly

vmarta committed:
Modal shows product set data

vmarta committed:
Parent product

vmarta committed:
Variant's product image shows up in add-to-cart modal

vmarta committed:
Rename variables for clarity

vmarta committed:
Fix linting error

vmarta committed:
Add horizontal rules to separate the parent and child items

vmarta committed:
Tweak whitespace

vmarta commented:
@bendvc all right, this PR is ready for another review. Aside from the UI changes, here are some main changes since your last review:
- Clearer argument name than `isSetProduct`: it's now `isProductPartOfSet`.
- Added the _parent_ product to the top of the page
- Added-to-cart modal now shows the correct product metadata and correct variant image too

vmarta commented on packages/template-retail-react-app/app/partials/product-view/index.jsx:
```diff
@@ -393,9 +405,15 @@ const ProductView = ({
-                        <Box display={['none', 'none', 'none', 'block']}>
-                            {renderActionButtons()}
-                        </Box>
+                        {!isProductASet && (
+                            <Box
+                                display={
+                                    isProductPartOfSet ? 'block' : ['none', 'none', 'none', 'block']
+                                }
+                            >
```
What do you think of `isProductASet` and `isProductPartOfSet`? How are they compared to the existing terminology seen in Business Manager and documentation (they use "product set" and "set product" to mean parent and child product, respectively)?

bendvc replied:
I assume that the reason they chose to `set product` is because it's short and its reversal of `product set` implicitly gives it a connection between the two types. 

That being said, if you read "set product" in isolation, likelihood that you would understand what it means is low. So although I normally like what old ocapi and their doc's, I think this isn't one of them. 

I'm not too concern either way, so you can do what you think is best on this topic, because we can change it since it's in the template. 

Another option (because I like to muddy the waters) would be something like `product.isSet` and `product.belongsToSet`

bendvc commented on packages/template-retail-react-app/app/hooks/use-product.js:
```diff
@@ -15,17 +15,18 @@ const OUT_OF_STOCK = 'OUT_OF_STOCK'
+    const variant = useVariant(product, isProductPartOfSet)
+    const variationParams = useVariationParams(product, isProductPartOfSet)
+    const variationAttributes = useVariationAttributes(product, isProductPartOfSet)
+    // console.log('--- variationAttributes', variationAttributes)
```
Don't forget to remove this 👍

vmarta committed:
Localize the Starting At label

vmarta committed:
Add todo
